poppy/sub_sampled_optics.py

```diff
@@ -212,157 +209,153 @@ def get_psfs(self):
         for i in range(self.x_apertures):
-            for j in  range(self.y_apertures):
+            for j in range(self.y_apertures):
                 self.wf_array[i][j] *= optic
         return
-    
 
     def get_centroids(self,
                       cent_function=measure_centroid,
-                      relativeto = 'origin',
+                      relativeto='origin',
                       asFITS=True,
                       **kwargs):
         """
         get centroid of intensity of each subwavefront
 
-        inputs: 
-        cent_function: function names 
+        inputs:
+        cent_function: function names
             see available funtions in .utils (if using a different one make sure to include it at top of this file as well)
         relative to: string
             can be either 'center' or 'origin' for default centroid function
         asFITS: bool
-            if True, treats subwavefronts as fits objects 
-        
-        returns: 
+            if True, treats subwavefronts as fits objects
+
+        returns:
         centroid_list: np array
             list of calculated centroids in shape (2, nx, ny)
-        note: if using wf_reconstruction method you need to either use 'center' here 
-        or provide a flat centroid list by propagating a flat wavefront through SHWFS 
+        note: if using wf_reconstruction method you need to either use 'center' here
+        or provide a flat centroid list by propagating a flat wavefront through SHWFS
         to get proper wf reconstruction
         recommended to keep to defaults since 'center' does weird stuff at edges of aperture
 
         """
+        print("Installed poppy in editable mode works!")
         _log.debug("Centroid function:"+str(cent_function))
         if self.input_wavefront is None:
             raise ValueError("No input wavefront found.")
         if not self._propagated_flag:
             _log.warn("Getting centroid without having propagated.")
-        self.centroid_list = np.zeros((2,self.x_apertures,self.y_apertures))
+        self.centroid_list = np.zeros((2,self.x_apertures, self.y_apertures))
         for i in range(self.x_apertures):
-            for j in  range(self.y_apertures):
+            for j in range(self.y_apertures):
                 sub_wf = self.wf_array[i][j]
                 if sub_wf.total_intensity == 0.0:
                     _log.warn("Setting centroid of aperture with no flux to NaN.")
-                    self.centroid_list[:,i,j]=(np.nan,np.nan)
+                    self.centroid_list[:, i, j] = (np.nan, np.nan)
                     continue
                 if asFITS:
-                    intensity_array=sub_wf.as_fits()
+                    intensity_array = sub_wf.as_fits()
                 else:
                     intensity_array = sub_wf.intensity
-                self.centroid_list[:,i,j] = cent_function(intensity_array,**kwargs, relativeto = relativeto)
```

Comment:
I'm not sure this is a bug since the asFITS case wouldn't lead to a centroid list if the function is only called inside the else statement - @remorgan123 do we have a test that includes this line?

Reply:
@douglase it looks like there is no unit test coverage at all for any of the sub sampled optics here - See the codecov report: 
https://codecov.io/gh/spacetelescope/poppy/compare/15a427893e59d22bb70172c660515edb6cebffd7...faa0187b70f4fa02a2bd783a54bc052ce6e43118/src/poppy/sub_sampled_optics.py?before=poppy/sub_sampled_optics.py

Reply:
him, not sure linking into codecov works well... if you go to the codecov link at the bottom of this PR, click to files, and browse to this file, then it shows 0% coverage.

Reply:
Sorry, I overlooked the asFITS case. I encountered the following error earlier while running [Shack Hartmann Wavefront Sensor Demo.ipynb](https://github.com/spacetelescope/poppy/blob/develop/notebooks/Shack%20Hartmann%20Wavefront%20Sensor%20Demo.ipynb) by changing the pixel_pitch from 2.2um to 5um. This leads to negative indices of `XLOC` and `YLOC` which I am not sure if it's intended by `fwcentroid`:

> ~/Projects/Python/Optics/poppy_sim/shwfs_demo.py in <module>
     88 shwfs.sample_wf(wf_flat)
     89 shwfs.get_psfs()
---> 90 flat_centroid_list = shwfs.get_centroids()
     91 
     92 #sample actual wf and propagate to detector:

> ~/Projects/Python/Modules/poppy/poppy/sub_sampled_optics.py in get_centroids(self, cent_function, relativeto, asFITS, **kwargs)
    272                 else:
    273                     intensity_array = sub_wf.intensity
--> 274                 self.centroid_list[:, i, j] = cent_function(intensity_array, **kwargs, relativeto=relativeto)
    275 
    276         self._centroided_flag = True

> ~/Projects/Python/Modules/poppy/poppy/utils.py in measure_centroid(HDUlist_or_filename, ext, slice, boxsize, verbose, units, relativeto, **kwargs)
   1024         image = image[slice, :, :]
   1025 
-> 1026     cent_of_mass = fwcentroid(image, halfwidth=boxsize, **kwargs)
   1027 
   1028     if verbose:

> ~/Projects/Python/Modules/poppy/poppy/fwcentroid.py in fwcentroid(image, checkbox, maxiterations, threshold, halfwidth, verbose)
    102             YLOC = j
    103             # print "  (%d, %d)" % (XLOC, YLOC)
--> 104             SUM += image[j, i]
    105             XSUM += XLOC * image[j, i]
    106             XSUM2 += XLOC ** 2 * image[j, i]

> IndexError: index 30 is out of bounds for axis 0 with size 30

My previous wrong intent bypassed the asFITS case and therefore hided the error but also led to no centroid estimation.

The error itself may not necessarily be a bug but could depend on a correct selections of the pixels per lenslet and `halfwidth`:
https://github.com/spacetelescope/poppy/blob/15a427893e59d22bb70172c660515edb6cebffd7/poppy/fwcentroid.py#L29

@mperrin  Could you please point me to some written materials on the two JWST materials so I can investigate further?
https://github.com/spacetelescope/poppy/blob/15a427893e59d22bb70172c660515edb6cebffd7/poppy/fwcentroid.py#L20

Reply:
yes, common error in my experience with fwcentroid() that usually requires adjusting the halfwidth size.

Reply:
Yes I agree the `fwcentroid` function isn't so sophisticated as it could be (by design... the flight version of this algorithm runs in a _very_ tiny and not so smart computer.) We could make the version here more intelligent about choosing good default parameter sizes, or at least giving a more informative error message that says the box size should be increased.